index.src.html

```diff
@@ -4179,10 +4205,36 @@
             [=combined buffer layout/allocationSize=].
     9. Return |combinedLayout|.
 
+: <dfn for=VideoFrame>Convert to RGB frame</dfn> (with |format|, |rect| and |colorSpace|)
+ :: 1. Let |canvas| be a new {{OffscreenCanvas}} constructed with
+        |rect|.|width| and |rect|.|height|.
```

Comment:
This isn't something that we can do in spec, using other specs directly like that. All the objects you use can have been overridden in the global this algorithm is running in.

We should use internal stops for canvas instead.

Reply:
I'm probably missing something here. Could you elaborate a bit?

Isn't it what we do in the Rendering section?

referring to [CanvasDrawImage](https://html.spec.whatwg.org/multipage/canvas.html#canvasdrawimage) [drawImage()](https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-drawimage)

Reply:
No, that's unrelated, we reference it in your link. What happens when `drawImage` is called with a `VideoFrame` is well-defined there. Here you're _using_ another Web Platform object from the innards of our spec.

e.g. what happens if in the document we're doing this conversion, I've maliciously done:

```js
window.OffscreenCanvas.prototype = function() { alert("what now"); };
```

before anything from our spec run?

Reply:
This algorithm does not refer to any particular js code. It explicitly says what object needs to be created by the user agent. That's why I don't think that global overrides should concern us here. With a similar approach half of this spec can be put into question with code like 

```
window.VideoFrame.prototype = function() { alert("what now"); };
```

Reply:
I added a note clarifying that UAs don't have to do it via literally creating an OffscreenCanvas as long as the result is the same. I hope this eliminates possible confusion. 

If you still think that the wording leaves room for misinterpretation, I'm very much open to suggestions how to rewrite it.

Reply:
This is not about "interpretation", it's simply not something we do in specs, and this needs to be rewritten with an abstract operation exported from the canvas code (if that's what you want to use), and calling into this abstract operation.

See https://github.com/whatwg/html/pull/5510#pullrequestreview-411744788, for a reference of this happening, and https://github.com/w3c/IntersectionObserver/issues/427 for the fix.

Reply:
Thanks for the links!

I've noticed that [drawImage()](https://html.spec.whatwg.org/multipage/canvas.html#drawing-images) (step 6) does not specify any details about the rendering (sampling, color conversion etc). It's only saying that "The user agent may use any filtering algorithm". 
So there isn't really an algorithm to extract and call. 
I will have to rewrite it in a more abstract manner, similar to the text for `drawImage()`

Reply:
We can add an export to something hand-wavy from canvas now, and merge what we want to do here, and then fix the canvas spec as well, your call.

Reply:
I [asked](https://github.com/whatwg/html/issues/10210) HTML spec owners about the best way to reuse canvas pixel format conversion algorithm.

Reply:
I've looked at the rest of the WebCodecs spec. It's pretty light on the details when it comes to format and color spaces conversions. (for example [Rendering](https://www.w3.org/TR/webcodecs/#videoframe-rendering) and [AudioData.copyTo](https://www.w3.org/TR/webcodecs/#dom-audiodata-copyto) )

So I rewrote "Convert to RGB frame" in the similar manner. 
Please take another look.

Reply:
@padenot gentle ping